From 4da7e4d4647003ac4b64db675fb288d66e9f9eda Mon Sep 17 00:00:00 2001
From: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Date: Fri, 23 Jul 2021 15:02:03 +0800
Subject: [PATCH] xhci: Show Zhaoxin XHCI root hub speed correctly
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 4da7e4d4647003ac4b64db675fb288d66e9f9eda


zhaoxin inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I40QDN
CVE: NA

----------------------------------------------------------------

Some Zhaoxin xHCI controllers follow usb3.1 spec,
but only support gen1 speed 5G. While in Linux kernel,
if xHCI suspport usb3.1,root hub speed will show on 10G.
To fix this issue, read usb speed ID supported by xHCI
to determine root hub speed.

Signed-off-by: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/usb/host/xhci.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index a834f3927bee..e9d67ee299d9 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -5186,6 +5186,7 @@ int xhci_gen_setup(struct usb_hcd *hcd, xhci_get_quirks_t get_quirks)
 	 */
 	struct device		*dev = hcd->self.sysdev;
 	unsigned int		minor_rev;
+	u8			i, j;
 	int			retval;
 
 	/* Accept arbitrarily long scatter-gather lists */
@@ -5240,6 +5241,24 @@ int xhci_gen_setup(struct usb_hcd *hcd, xhci_get_quirks_t get_quirks)
 			hcd->self.root_hub->speed = USB_SPEED_SUPER_PLUS;
 			break;
 		}
+
+		/* usb3.1 has gen1 and gen2, Some zx's xHCI controller that follow usb3.1 spec
+		 * but only support gen1
+		 */
+		if (xhci->quirks & XHCI_ZHAOXIN_HOST) {
+			minor_rev = 0;
+			for (j = 0; j < xhci->num_port_caps; j++) {
+				for (i = 0; i < xhci->port_caps[j].psi_count; i++) {
+					if (XHCI_EXT_PORT_PSIV(xhci->port_caps[j].psi[i]) >= 5)
+						minor_rev = 1;
+				}
+				if (minor_rev != 1) {
+					hcd->speed = HCD_USB3;
+					hcd->self.root_hub->speed = USB_SPEED_SUPER;
+				}
+			}
+		}
+
 		xhci_info(xhci, "Host supports USB 3.%x %sSuperSpeed\n",
 			  minor_rev,
 			  minor_rev ? "Enhanced " : "");
-- 
2.26.2

